cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,231 @@
+		podIPMap[podList.Items[i].Status.PodIP] = &podList.Items[i]
+	}
+
+	// Migrate IPIP tunnel address.
+	if node, err = calicoClient.Nodes().Get(ctxt, nodename, options.GetOptions{}); err != nil {
+		msg = "failed to get node resource"
+		return
+	}
+	ipipTunnelAddr := cnet.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+	if err = calicoClient.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+		IP:       *ipipTunnelAddr,
+		Hostname: nodename,
+	}); err != nil {
+		msg = "failed to get add IPIP tunnel addr " + node.Spec.BGP.IPv4IPIPTunnelAddr
+		return
+	}
+
+	// Read in all the files in the host-local directory.
+	if files, err = ioutil.ReadDir(ipAllocPath); err != nil {
+		msg = "failed to read path " + ipAllocPath
+		return
+	}
+
+	// For each file, convert it into an IP allocation and then delete the file.
+	for _, f := range files {
+		fname := path.Join(ipAllocPath, f.Name())
+
+		// Delete and skip the last reserved IP.
+		if strings.Contains(f.Name(), "last") {
+			if err = os.Remove(fname); err != nil {
+				msg = "failed to remove file " + fname
+				return
+			}
+			continue
+		}
+
+		// The name of the file is its IP address.
+		if ip, _, err = cnet.ParseCIDR(fmt.Sprintf("%s/32", f.Name())); err != nil {
+			msg = "failed to parse IP " + f.Name()
+			return
+		}
+
+		// The contents are the container ID.
+		if b, err = ioutil.ReadFile(fname); err != nil {
+			msg = "failed to read file " + fname
+			return
+		}
+
+		// Get the pod resource associated with the IP.
+		pod, ok := podIPMap[f.Name()]
+		if !ok {
+			log.WithField("ip", f.Name()).Info("pod not found for IP, deleting and continuing...")
+			if err = os.Remove(fname); err != nil {
+				msg = "failed to remove file " + fname
+				return
+			}
+			continue
+		}
+
+		// Store allocation to datastore
+		handleID := fmt.Sprintf("%s.%s", "k8s-pod-network", string(b))
+		if err = calicoClient.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+			IP:       *ip,
+			HandleID: &handleID,
+			Hostname: nodename,
+			Attrs: map[string]string{
+				"node":      nodename,
+				"pod":       pod.Name,
+				"namespace": pod.Namespace,
+			},
+		}); err != nil {
+			msg = "failed to assign IP to calico backend"
+			return
+		}
+
+		// Delete the file
+		if err = os.Remove(fname); err != nil {
+			msg = "failed to remove file " + fname
+			return
+		}
+	}
+
+	// Delete the directory
+	if err = os.Remove(ipAllocPath); err != nil {
+		msg = "failed to remove directory " + ipAllocPath
+		return
+	}
+
+	// Check DaemonSet rolling update status to see if this is the last node to update.
+	// If so, re-enable auto-allocation of blocks
+	if calicoDS, err = k8sClient.ExtensionsV1beta1().DaemonSets("kube-system").Get("calico-node", v1.GetOptions{}); err != nil {
+		msg = "failed to get daemonsets/kube-system.calico-node"
+		return
+	}
+
+	// UpdatedNumberScheduled tracks the total number of nodes that are running updated daemon pod.
+	if calicoDS.Status.UpdatedNumberScheduled == calicoDS.Status.DesiredNumberScheduled-1 {
```

Comment:
Wonder if we should look at `maxUnavailable` here.

e.g., if someone decides to upgrade 5 pods at a time, then it's possible none of them will detect they are the last one. 

Probably better if all of them think they are the last one.